File: 003_longest_substring_without_replacing_characters/hash.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        """
        Complexity: O(n)
        Space: O(n)
        use hash to store used char
        """
        max_length = 0
        start = 0
        used_char = {}
        for i, c in enumerate(s):
            if c in used_char and start <= used_char[c]:
                start = used_char[c] + 1 # start from the next char of the last occurrence
            else:
                max_length = max(max_length, i-start+1)
            used_char[c] = i
                
        return max_length

File: 003_longest_substring_without_replacing_characters/hash.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0;
        int max_length = 0;
        std::unordered_map<char, int>used_char;
        // std::vector<int>used_char(256, -1)
        // if char using ASCII, could use this vector as the hashmap
        // need to make sure start <= used_char[[s[i]]]
        // since the idx used_char[s[i]] may be smaller than the start due to other repeating chars
        for(int i = 0; i<s.length(); ++i){
            if(used_char.find(s[i]) != used_char.end() && start <= used_char[s[i]]){
                start = used_char[s[i]] + 1;
            }
            else{
                max_length = std::max(max_length, i-start+1);
            }
            used_char[s[i]] = i;
        }
        return max_length;
    }
};
